File: scripts/nanobanana.mjs
#!/usr/bin/env node
/*
 * nanobanana.mjs — directed image generation with Google's Gemini image models,
 * using plain Node (v18+ global fetch). A drop-in for nanobanana.py when Python
 * isn't installed on Windows.
 *
 *   node scripts/nanobanana.mjs "a tin of dog food, studio lit" --out public/images/x.png
 *
 * Flags: --out <path>  --model <id>  --aspect 1:1  --size 1K  --ref <img> [--ref ...]
 * Key:   reads GEMINI_API_KEY, else scripts/gemini-key.txt
 */
import { readFileSync, writeFileSync, existsSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'

const __dirname = dirname(fileURLToPath(import.meta.url))
const DEFAULT_MODEL = 'gemini-3.1-flash-image'

function getKey() {
  if (process.env.GEMINI_API_KEY) return process.env.GEMINI_API_KEY.trim()
  const keyFile = join(__dirname, 'gemini-key.txt')
  if (existsSync(keyFile)) return readFileSync(keyFile, 'utf8').trim()
  throw new Error('No GEMINI_API_KEY env var and no scripts/gemini-key.txt')
}

function parseArgs(argv) {
  const a = { prompt: null, out: 'output.png', model: DEFAULT_MODEL, aspect: null, size: null, ref: [] }
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i]
    if (t === '--out') a.out = argv[++i]
    else if (t === '--model') a.model = argv[++i]
    else if (t === '--aspect') a.aspect = argv[++i]
    else if (t === '--size') a.size = argv[++i]
    else if (t === '--ref') a.ref.push(argv[++i])
    else if (!a.prompt) a.prompt = t
  }
  return a
}

function guessMime(path) {
  const p = path.toLowerCase()
  if (p.endsWith('.jpg') || p.endsWith('.jpeg')) return 'image/jpeg'
  if (p.endsWith('.webp')) return 'image/webp'
  return 'image/png'
}

export async function generate({ prompt, out, model = DEFAULT_MODEL, aspect, size, ref = [] }) {
  const key = getKey()
  const parts = [{ text: prompt }]
  for (const r of ref) {
    const buf = readFileSync(r)
    parts.push({ inlineData: { mimeType: guessMime(r), data: buf.toString('base64') } })
  }
  const imageConfig = {}
  if (aspect) imageConfig.aspectRatio = aspect
  if (size) imageConfig.imageSize = size

  const body = {
    contents: [{ parts }],
    generationConfig: {
      responseModalities: ['TEXT', 'IMAGE'],
      ...(Object.keys(imageConfig).length ? { imageConfig } : {}),
    },
  }

  const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent`
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'x-goog-api-key': key },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const txt = await res.text()
    throw new Error(`HTTP ${res.status}: ${txt.slice(0, 800)}`)
  }
  const json = await res.json()
  const cand = json?.candidates?.[0]
  const respParts = cand?.content?.parts || []
  let saved = null
  const texts = []
  for (const part of respParts) {
    if (part.text) texts.push(part.text.trim())
    else if (part.inlineData?.data) {
      writeFileSync(out, Buffer.from(part.inlineData.data, 'base64'))
      saved = out
    }
  }
  return { saved, texts, finishReason: cand?.finishReason, raw: json }
}

// CLI
if (import.meta.url === `file://${process.argv[1]}` || process.argv[1]?.endsWith('nanobanana.mjs')) {
  const args = parseArgs(process.argv.slice(2))
  if (!args.prompt) {
    console.error('Give a prompt. e.g. node scripts/nanobanana.mjs "a dog" --out out.png --aspect 1:1 --size 1K')
    process.exit(1)
  }
  try {
    const { saved, texts, finishReason } = await generate(args)
    if (texts.length) console.log('[model]', texts.join(' '))
    if (saved) console.log('Saved:', saved)
    else console.log('No image returned. finishReason:', finishReason)
  } catch (e) {
    console.error('ERROR:', e.message)
    process.exit(1)
  }
}

File: scripts/nanobanana.py
#!/usr/bin/env python3
"""
nanobanana.py — directed image generation with Google's Gemini image models.

Two ways to drive it:
  1. One-shot (great for an agent like Claude Code to call, inspect, then re-call):
       python nanobanana.py "a tin of Buggly Farms cricket protein, studio lit" --out can.png
  2. Interactive chat (you steer it turn by turn; context is preserved so you can
     say "make the label bigger" and it edits the SAME image):
       python nanobanana.py --chat

Setup:
  pip install google-genai pillow
  export GEMINI_API_KEY="your_key_from_aistudio.google.com/apikey"

Models (--model):
  gemini-3.1-flash-image   Nano Banana 2  (default; fast, 512/1K/2K/4K, up to 14 ref images)
  gemini-3-pro-image       Nano Banana Pro (highest quality, best text rendering)
  gemini-2.5-flash-image   Nano Banana    (cheapest)
"""

import argparse
import os
import sys
from pathlib import Path

try:
    from google import genai
    from google.genai import types
    from PIL import Image
except ImportError:
    sys.exit("Missing deps. Run: pip install google-genai pillow")

DEFAULT_MODEL = "gemini-3.1-flash-image"


def build_config(aspect: str | None, size: str | None) -> types.GenerateContentConfig:
    img_cfg = None
    if aspect or size:
        kwargs = {}
        if aspect:
            kwargs["aspect_ratio"] = aspect          # e.g. 1:1, 16:9, 9:16, 4:5, 21:9
        if size:
            kwargs["image_size"] = size              # 512, 1K, 2K, 4K  (uppercase K)
        img_cfg = types.ImageConfig(**kwargs)
    return types.GenerateContentConfig(
        response_modalities=["TEXT", "IMAGE"],
        image_config=img_cfg,
    )


def save_images(response, out_base: Path) -> list[Path]:
    """Walk response parts, save every image, print any text the model returned."""
    saved = []
    idx = 0
    for part in response.parts:
        if getattr(part, "text", None):
            print(f"[model] {part.text.strip()}")
        elif getattr(part, "inline_data", None) is not None:
            img = part.as_image()
            path = out_base if idx == 0 else out_base.with_stem(f"{out_base.stem}_{idx}")
            img.save(path)
            saved.append(path)
            idx += 1
    return saved


def one_shot(client, args):
    contents = [args.prompt]
    for ref in args.ref or []:
        contents.append(Image.open(ref))     # reference / image-to-edit inputs
    resp = client.models.generate_content(
        model=args.model,
        contents=contents,
        config=build_config(args.aspect, args.size),
    )
    saved = save_images(resp, Path(args.out))
    if saved:
        print("Saved:", ", ".join(str(p) for p in saved))
    else:
        print("No image returned. Model said:",
              getattr(resp, "text", "") or "(nothing — possibly blocked by safety filters)")


def chat_loop(client, args):
    cfg = build_config(args.aspect, args.size)
    chat = client.chats.create(model=args.model, config=cfg)
    n = 0
    print(f"Interactive mode — model={args.model}. "
          "Type a prompt to generate, then refine ('make it warmer', 'add fog').")
    print("Commands: /new (reset context)  /model <name>  /quit\n")
    while True:
        try:
            msg = input("> ").strip()
        except (EOFError, KeyboardInterrupt):
            print()
            break
        if not msg:
            continue
        if msg == "/quit":
            break
        if msg == "/new":
            chat = client.chats.create(model=args.model, config=cfg)
            print("[context reset]")
            continue
        if msg.startswith("/model "):
            args.model = msg.split(" ", 1)[1].strip()
            chat = client.chats.create(model=args.model, config=cfg)
            print(f"[model -> {args.model}, context reset]")
            continue
        resp = chat.send_message(msg)
        saved = save_images(resp, Path(args.out).with_stem(f"{Path(args.out).stem}_{n:02d}"))
        if saved:
            print("Saved:", ", ".join(str(p) for p in saved))
        n += 1


def main():
    p = argparse.ArgumentParser(description="Directed Gemini image generation.")
    p.add_argument("prompt", nargs="?", help="Text prompt (omit if using --chat).")
    p.add_argument("--chat", action="store_true", help="Interactive multi-turn editing.")
    p.add_argument("--model", default=DEFAULT_MODEL)
    p.add_argument("--out", default="output.png", help="Output filename / base.")
    p.add_argument("--ref", nargs="*", help="Reference image path(s) to edit or guide from.")
    p.add_argument("--aspect", help="1:1, 16:9, 9:16, 4:5, 3:2, 21:9, etc.")
    p.add_argument("--size", help="512, 1K, 2K, 4K (uppercase K).")
    args = p.parse_args()

    if not os.environ.get("GEMINI_API_KEY"):
        sys.exit("Set GEMINI_API_KEY first (get one at aistudio.google.com/apikey).")

    client = genai.Client()   # reads GEMINI_API_KEY automatically

    if args.chat:
        chat_loop(client, args)
    elif args.prompt:
        one_shot(client, args)
    else:
        p.error("Give a prompt, or use --chat.")


if __name__ == "__main__":
    main()
